[#7] Spring Sercurity 적용

jjeda opened the pull request:
인증 & 인가를 위한 Spirng Security 적용
- spring security 적용
- AuthServer 설정

jjeda committed:
[#7] spring security 적용

- Spring security oAuth Dependency 추가
- Security config 클래스 추가
- Account 객체 내 userName 필드 -> nickname 으로 변경
- AccountService 객체 -> UserDetailService Implements

jjeda committed:
[#7] AuthServer 설정

- 인증 & 인가를 위한 Auth 서버 설정

f-lab-dev commented on src/main/java/me/jjeda/mall/accounts/configs/SecurityConfig.java:
```diff
@@ -0,0 +1,70 @@
+    protected void configure(AuthenticationManagerBuilder auth) throws Exception {
+        auth
+                .userDetailsService(accountService)
+                .passwordEncoder(this.passwordEncoder());
```
이렇게 수동으로 객체를 생성해주면 Bean과 별도의 객체가 생성되어 2개의 객체가 공존하게되고, 관리가 어렵게 됩니다.
오버라이드 메소드라 이렇게 하신 것 같지만 Config 파일을 분리하는 등 객체를 주입받을 수 있는 방식으로 바꾸시는게 좋을 것 같습니다

jjeda replied:
이부분이 잘 이해가 가질 않는데요..
userDeatailsService를 제가 구현한 accountService로
passwordEncoder도 제가 원하는 passwordEncoder로 설정하는 메소드로 이해를 했습니다.
2개의 객체가 공존하게 된다는 말씀을 이해하지 못하였습니다..

f-lab-dev replied:
`passwordEncoder` 메소드에 `@Bean` 어노테이션을 정의해두셨는데요. 이렇게 하면 스프링이 관리하는 passwordEncoder 객체가 하나가 생성이되고요.
configure 메소드에서 이곳처럼 this.passwordEncoder()를 직접 호출하셔서 별도의 객체를 생성하면 총 2개의 객체가 생기게 됩니다.
`@Bean` 어노테이션을 달아주는 이유는 이 객체를 스프링이 관리하게 하겠다는 선언입니다.
스프링에게 객체 생성방법을 알려주는 용도로 만든 것인데 직접 호출해서 객체를 생성하면 원래의 목적에 어긋나게 사용하는 것이 됩니다.

jjeda replied:
와.. 이론적으로 알고는 있었지만 머리속에 전혀 신경도 못쓰고 있었던 부분입니다..
코드 바로 수정하겠습니다~

f-lab-dev commented on src/main/java/me/jjeda/mall/accounts/Service/AccountService.java:
```diff
@@ -15,13 +19,17 @@
 @Service
 @AllArgsConstructor
 @Transactional
```
전역 Transactional은 좋지 않아보입니다.

f-lab-dev commented on src/main/java/me/jjeda/mall/accounts/configs/AuthServerConfig.java:
```diff
@@ -0,0 +1,56 @@
+@AllArgsConstructor
+public class AuthServerConfig extends AuthorizationServerConfigurerAdapter {
+
+    private PasswordEncoder passwordEncoder;
```
사실 스프링이 잘 주입해주긴 합니다만. 주입받을 것과 주입받지 않을 것을 구분하는 것도 중요합니다.
일괄적으로 주입받기보다는 명시적으로 `@AutoWired`나 `@Resource`를 사용하여 주입받을 객체를 명시적으로 처리해주는 것이 좋습니다.

굳이 setter가 필요하면 `@Setter`를 사용해 생성해주시면 좋고요~

jjeda replied:
의존성 주입방법 3가지 1. 필드 2. Setter 3. 생성자
3가지 방법중에 Spring이 추천하는 방법은 3번 생성자로 알고있습니다! 스프링 버전이 올라가면서 생성자에 
`@Autowired`를 생략해도 되는 것으로 알고있습니다 필드에 애노태이션을 붙이는 것보다 생성자를 만들어주는 방법은 어떤가요? (`@AllArgsConstructor`가 아닌)

f-lab-dev replied:
좋은 생각입니다~

jjeda committed:
[#7] Auth & Resource Server configuration

- AuthServer 설정
- ResourceServer 설정
- SecurityConfig -> 공통설정
- RedisTokenStore 관련 코드 추가
- 권한타입 String -> Role 변경

f-lab-dev commented on src/main/java/me/jjeda/mall/accounts/Service/AccountService.java:
```diff
@@ -2,37 +2,52 @@
     public Account saveAccount(AccountDto dto) {
         Account account = dto.toEntity();
-        account.setPassword(passwordEncoder.encode(dto.getPassword()));
+        account.setPassword(this.passwordEncoder().encode(dto.getPassword()));
```
setPassword를 할 때마다 passwordEncoder 객체를 새로 생성하게 하는 것 같은데요.
이렇게 하면 100만번 호출하면 객체 100만개가 생성됩니다.
`@Bean`으로 정의해두셨으니 스프링에서 관리하는 객체 하나만 생성해서 주입받아 사용하면 되는데요~

이 쪽이 아직 잘 이해가 안돼시면 오프라인 멘토링때 자세히 설명드리겠습니다~

f-lab-dev replied:
[참고] 객체를 생성하고 관리하는 비용은 비쌉니다. 생성하는 비용뿐만 아니라 GC시에 살아있는 객체를 식별하고, 청소하는 등 여러 곳에서 부담을 주어서 stateless한 객체는 싱글톤으로 만들어주는 것이 좋습니다.
스프링의 빈들도 싱글톤으로 관리되고요.

jjeda committed:
[#7] 인증 &인가 관련 수정

- PassEncoder @bean을 위한 공통 config 클래스 생성
- 명시성을 위해 빈주입시 @AllArgsConstructor -> 생성자

jjeda committed:
[#10] Validation 추가 및 Authorization

- AccountDto 객체에 Validation 추가
- Validation 검증을 위한 테스트코드 추가
- 관리자와 유저 account 컨트롤러 분리
- Role에 따른 Authrization 범위 설정
- Authrization 테스트 코드 추가

jjeda committed:
[#10] AccountStatus 추가 & 사용자 로직 수정

- isDeleted Flag 삭제 -> 상태를 다양화 할 수 있는 AccountStatus enum 추가
- 상태전이 코드 추가
- 사용자가 @Pathvariable을 통해 endpoint로 접근하는 것이 아닌 현재사용자 정보를 받아와 접근하게 수정
- 기타 : import에 check style 적용, Appconfig 이름 역할에 맞게 수정

jjeda committed:
[#10] Current User 로직 수정

- AccountAdapter 클래스 구현 ( extends User)
- 어댑터 패턴 적용

jjeda committed:
[#10] 현재 사용자 로직수정

- expression = account SPEL을 통해 간결하게 변경
- 간결함, 가독성을 위해 CurrentUser 애노태이션 정의

jjeda committed:
[#16] Account 관련 수정 및 개선

- unused import 삭제
- 변수명 및 매서드명 수정

jjeda committed:
Merge pull request #18 from f-lab-edu/feature/16

[#16] Account 관련 수정 및 개선

jjeda committed:
Merge pull request #14 from f-lab-edu/feature/10

[#10] Validation 추가 및 Authorization
